### PR TITLE
chore(ui): migrate Badge component to Typescript

### PR DESCRIPTION
**Pull request by guoda-puidokaite**
# Summary

- Migrated the `Badge` component to `TypeScript`, enhancing code quality, maintainability, and scalability.
- Utilised static typing to minimize runtime errors and improve the development experience.
- Refactored the component for cleaner, more maintainable code.

# Changes Made

**TypeScript Migration**:

- Converted `Badge` component files to `TypeScript`.
- Added comprehensive type annotations.

**Refactoring**:

- Made necessary adjustments to address TypeScript-detected issues.
- Further refactored for cleaner and more understandable code.

**Bug Fix:**

- Removed the critical variant, as discussed [here](https://github.com/cloudoperators/juno/issues/213#issuecomment-2413067115).

**Testing**:

- Organised tests by what they check for better readability and maintenance.
- Added edge case coverage for: 
  - default rendering with no props
  - additional CSS classes
  - handling `undefined` icons.
- Verified component behaviour using `Storybook`.
- Updated tests to use `vitest`.

**Cleanup:**

- Removed deprecated code. It wasn't needed for any `.js` components.

**Documentation:**

- Updated `Storybook` documentation.

Screenshot: Deprecated `.js` code on the left, migrated `.ts` code on the right (**see notes**)
Screenshot 2024-10-14 at 09 39 54

**Release:**

- Generated a `changeset` file, mentioning the migration and removal of the critical variant.

# Related Issues

- https://github.com/cloudoperators/juno/issues/213

# Testing Instructions

1. Run `Storybook`
2. Run `Badge` tests

# Checklist

- [x] I have performed a self-review of my code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have made corresponding changes to the documentation (if applicable).
- [x] My changes generate no new warnings or errors.

### CODE REVIEW

**Review by andypf — DISMISSED** [review]
Great work, Guoda. I like that you added descriptions in storybook. It makes the properties even more understandable. Very well done!

**Comment by guoda-puidokaite** [comment]
Aw! Thank You so much for the review, @andypf! 💪

File: packages/ui-components/src/components/Badge/Badge.test.tsx [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+  test("renders a badge with children", () => {
+    render(<Badge data-testid="badge">Children inside</Badge>)
+    expect(screen.getByTestId("badge")).toBeInTheDocument()
+    expect(screen.getByTestId("badge")).toHaveTextContent("Children inside")
+  })
+
+  // VARIANTS
+
+  const variants: VariantType[] = ["info", "success", "warning", "danger", "error", "critical"]
+  variants.forEach((variant) => {
```
Comment (barsukov) [inline packages/ui-components/src/components/Badge/Badge.test.tsx:40]: I would suggest to not reuse the tests here with forEach, duplications in tests are even suggestible so lets discuss it too.

File: packages/ui-components/src/components/Badge/Badge.test.tsx [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  // ICONS
+
+  test("renders with default icon based on variant", () => {
+    render(<Badge icon={true} />)
+    expect(screen.getByRole("img")).toBeInTheDocument()
+    expect(screen.getByRole("img")).toHaveClass("juno-icon-default")
+  })
+
+  const icons: KnownIcons[] = ["comment", "deleteForever"]
+  icons.forEach((icon) => {
```
Comment (barsukov) [inline packages/ui-components/src/components/Badge/Badge.test.tsx:57]: same here

File: packages/ui-components/src/components/Badge/Badge.types.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
+/*
+ * SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Juno contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { KnownIcons } from "../Icon/Icon.component"
+
+// To Do: Externalise? Similar is used in Message component
+export type VariantType = "default" | "info" | "success" | "warning" | "danger" | "error" | "critical"
+export interface BadgeProps {
```
Comment (barsukov) [inline packages/ui-components/src/components/Badge/Badge.types.ts:10]: So you put types separately from the file, everywhere we do other way around. I know it is a best practises what you do but it is really small component, I would suggest to keep it in one file, and when it is a complicated type we could separated.

File: packages/ui-components/src/components/Badge/Badge.types.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
+/*
+ * SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Juno contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { KnownIcons } from "../Icon/Icon.component"
+
+// To Do: Externalise? Similar is used in Message component
```
Comment (barsukov) [inline packages/ui-components/src/components/Badge/Badge.types.ts:8]: No, we should not do it please check this article https://sandimetz.com/blog/2016/1/20/the-wrong-abstraction#:~:text=Once%20an%20abstraction%20is%20proved%20wrong%20the%20best,abandon%20the%20wrong%20abstraction%20sooner%20rather%20than%20later.

File: packages/ui-components/src/components/Badge/Badge.component.tsx [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+const getVariantStyle = (variant: VariantType): string => badgeVariantStyles[variant] || badgeVariantStyles.default
+
+// ICON UTILS
+
+const iconStyles = "jn-mr-1 jn-items-center"
+
+// Get icon based on the provided variant or icon string
+const getIcon = (icon: boolean | KnownIcons | undefined, variant: VariantType): KnownIcons | undefined => {
```
Comment (barsukov) [inline packages/ui-components/src/components/Badge/Badge.component.tsx:40]: lets discuss this one on personal call, I have few suggestions :)

**Review by barsukov — CHANGES_REQUESTED** [review]
I really like this PR, and the ideas that you have here! But I would like to make some improvements that can levitate it even further :) Let's have a call together and we will find a common ground on some discussion points that I made. Overall good job 👍

File: packages/ui-components/src/components/Badge/Badge.component.tsx [hunk 6]
```diff
@@ -30,27 +60,24 @@ const badgeVariantStyles = {
...
-// ICON UTILS
+const getVariantStyle = (variant: BadgeVariantType): string => badgeVariantStyles[variant] || badgeVariantStyles.default
 
 const iconStyles = "jn-mr-1 jn-items-center"
 
 // Get icon based on the provided variant or icon string
-const getIcon = (icon: boolean | KnownIcons | undefined, variant: VariantType): KnownIcons | undefined => {
+const getIcon = (icon: boolean | KnownIcons | undefined, variant: BadgeVariantType): KnownIcons | undefined => {
   if (typeof icon === "string" && isValidIcon(icon)) return icon
   if (icon === true) return variant as KnownIcons
```
Comment (barsukov) [inline packages/ui-components/src/components/Badge/Badge.component.tsx:70]: remember my code that I told you to construct the type? We should change this line, I already sent you a snippets in slack yesterday.

Comment (guoda-puidokaite) [reply packages/ui-components/src/components/Badge/Badge.component.tsx]: Thanks, @barsukov. I was waiting for a decision on the critical variant, as [here](https://github.com/cloudoperators/juno/pull/514#discussion_r1800561824). 
It was decided that the `critical` variant shouldn't exist at all, so we can remove it all together and also just remove the forced casting: `variant as KnownIcons`! Wooo. 🚀
I think that we don't need the solution from yesterday? But let me know if I missed something.

File: packages/ui-components/src/components/Badge/Badge.component.tsx (hunk 5 above)
Comment (guoda-puidokaite) [reply packages/ui-components/src/components/Badge/Badge.component.tsx]: Resolution described [here](https://github.com/cloudoperators/juno/pull/514#discussion_r1800619164).

**Review by barsukov — DISMISSED** [review]
🎆 Amazing! 💯

**Review by edda — APPROVED** [review]
Very thorough, thank you!

**Review by andypf — APPROVED** [review]
